Adding Vertex AI Imagen models support

Marzouki-Sami opened the pull request:
Closes gh-2133

Signed-off-by: Marzouki-Sami samymarzouki502@gmail.com

Thank you for taking time to contribute this pull request!
You might have already read the [contributor guide][1], but as a reminder, please make sure to:

* Sign the [contributor license agreement](https://cla.pivotal.io/sign/spring)
* Rebase your changes on the latest `main` branch and squash your commits
* Add/Update unit tests as needed
* Run a build and make sure all tests pass prior to submission

Marzouki-Sami committed:
Adding Vertex AI Imagen models support

Closes gh-2133

Signed-off-by: Marzouki-Sami samymarzouki502@gmail.com
Signed-off-by: Marzouki-Sami <samymarzouki502@gmail.com>

markpollack commented:
@ddobrin can you please take a look?

turchinc commented:
Any update here? I am also interested in this feature.

byungjunn commented:
I am currently using this feature as a custom-made version. I hope it gets updated soon. If help is needed, I would be willing to participate.

ilayaperumalg commented:
@byungjunn Thanks! Any help reviewing the PR is much appreciated.

byungjunn commented on models/spring-ai-vertex-ai-imagen/src/main/java/org/springframework/ai/vertexai/imagen/VertexAiImagenImageModel.java:
```diff
@@ -0,0 +1,256 @@
+		var observationContext = ImageModelObservationContext.builder()
+				.imagePrompt(imagePrompt)
+				.provider(AiProvider.VERTEX_AI.value())
+				.requestOptions(finalOptions)
```
[requestOptions](https://github.com/spring-projects/spring-ai/blob/83294023cd4946c46b42df01b686bcc70418ea70/spring-ai-core/src/main/java/org/springframework/ai/image/observation/ImageModelObservationContext.java#L86) is deprecated.

How about merging the default options in the mergedOptions method and then creating a new imagePrompt with finalOptions?

byungjunn commented on spring-ai-spring-boot-starters/spring-ai-starter-vertex-ai-imagen/pom.xml:
```diff
@@ -0,0 +1,61 @@
+	<parent>
+		<groupId>org.springframework.ai</groupId>
+		<artifactId>spring-ai</artifactId>
+		<version>1.0.0-M5</version>
```
```suggestion
		<version>1.0.0-SNAPSHOT</version>
```

byungjunn commented on spring-ai-spring-boot-starters/spring-ai-starter-vertex-ai-imagen/pom.xml:
```diff
@@ -0,0 +1,61 @@
+		<relativePath>../../pom.xml</relativePath>
+	</parent>
+
+	<artifactId>spring-ai-vertex-ai-imagen-spring-boot-starter</artifactId>
```
```suggestion
	<artifactId>spring-ai-starter-model-vertex-ai-imagen</artifactId>
```

byungjunn commented on spring-ai-spring-boot-starters/spring-ai-starter-vertex-ai-imagen/pom.xml:
```diff
@@ -0,0 +1,61 @@
+
+	<artifactId>spring-ai-vertex-ai-imagen-spring-boot-starter</artifactId>
+	<packaging>jar</packaging>
+	<name>Spring AI Starter - VertexAI Imagen</name>
```
```suggestion
    <name>Spring AI Starter - VertexAI Gemini</name>
    <description>Spring AI Vertex Gemini AI Spring Boot Starter</description>
```

byungjunn commented on spring-ai-spring-boot-starters/spring-ai-starter-vertex-ai-imagen/pom.xml:
```diff
@@ -0,0 +1,61 @@
+
+		<dependency>
+			<groupId>org.springframework.ai</groupId>
+			<artifactId>spring-ai-spring-boot-autoconfigure</artifactId>
```
```suggestion
			<artifactId>spring-ai-autoconfigure-model-vertex-ai</artifactId>
```

byungjunn commented:
@Marzouki-Sami 
https://spring.io/blog/2025/03/25/spring-ai-update-to-snapshots
After the update, there have been many changes, so I think you need to rebase onto the main branch.

Marzouki-Sami committed:
Update spring-ai-spring-boot-starters/spring-ai-starter-vertex-ai-imagen/pom.xml

Updating the version.

Co-authored-by: ByungJun <kbj7227@naver.com>
Signed-off-by: Sami Marzouki <73302274+Marzouki-Sami@users.noreply.github.com>

Marzouki-Sami committed:
Update spring-ai-spring-boot-starters/spring-ai-starter-vertex-ai-imagen/pom.xml

changing artifactId

Co-authored-by: ByungJun <kbj7227@naver.com>
Signed-off-by: Sami Marzouki <73302274+Marzouki-Sami@users.noreply.github.com>

Marzouki-Sami committed:
Update spring-ai-spring-boot-starters/spring-ai-starter-vertex-ai-imagen/pom.xml

removing unnecessary spaces

Co-authored-by: ByungJun <kbj7227@naver.com>
Signed-off-by: Sami Marzouki <73302274+Marzouki-Sami@users.noreply.github.com>

Marzouki-Sami committed:
updating artifactId

Co-authored-by: ByungJun <kbj7227@naver.com>
Signed-off-by: Sami Marzouki <73302274+Marzouki-Sami@users.noreply.github.com>

sami-marzouki-BI committed:
resolving conflict

Signed-off-by: Marzouki-Sami <samymarzouki502@gmail.com>

Marzouki-Sami committed:
updating to version 1.1.0-SNAPSHOT
Signed-off-by: Marzouki-Sami <samymarzouki502@gmail.com>

Marzouki-Sami commented:
@byungjunn @ilayaperumalg @markpollack i had some issues with resolving the conflict and to resolve the conflict i had to remove the commits, doing that closed this PR is it possible to reopen it, if not i will create a new PR and link it to this one.

Marzouki-Sami committed:
Merge branch 'main' into gh-2133

Signed-off-by: Sami Marzouki <73302274+Marzouki-Sami@users.noreply.github.com>

Marzouki-Sami commented:
Any updates?

Marzouki-Sami committed:
Update VertexAiImagenImageGenerationMetadata.java

switching constructor params to match only call in VertexAiImagenImageModel.java

Signed-off-by: Marzouki-Sami samymarzouki502@gmail.com
Signed-off-by: Marzouki-Sami <samymarzouki502@gmail.com>

Marzouki-Sami commented:
Just a minor change added in the last commit (f341567d9aaa0f618fecbcab0e260da69ec12bb6). Switching the constructor params ( mimetype and model) of `VertexAiImagenImageGenerationMetadata` to be in the same order of usage in `VertexAiImagenImageModel` ( line 177)

Marzouki-Sami committed:
adding the new models + setting older models as deprecated + adding 3 missing imageOptions
Author: Marzouki-Sami <samymarzouki502@gmail.com>
Signed-off-by: Marzouki-Sami samymarzouki502@gmail.com
Signed-off-by: Marzouki-Sami <samymarzouki502@gmail.com>

Marzouki-Sami commented:
@markpollack @ilayaperumalg @ddobrin @byungjunn Any updates? If not, will it take much longer?

Marzouki-Sami committed:
Merge branch 'main' into gh-2133

Signed-off-by: Sami Marzouki <73302274+Marzouki-Sami@users.noreply.github.com>

Marzouki-Sami commented:
@markpollack @ilayaperumalg @ddobrin This pull request has been open for over a year without a response to previous mentions. I understand the team may be busy, so if you don't have the capacity to review this at the moment, could you please delegate the review to another team member with write access?

Marzouki-Sami committed:
Merge branch 'main' into gh-2133

Signed-off-by: Sami Marzouki <73302274+Marzouki-Sami@users.noreply.github.com>

Marzouki-Sami commented:
@markpollack @ilayaperumalg @ddobrin Another conflict occurred, and there has still been no response to any of the previous comments.

OreOreDa commented:
I created another clean PR, in order to solve the issue : https://github.com/spring-projects/spring-ai/issues/2133
Please have a look

ddobrin commented:
Hi @OreOreDa 
please note that the Imagen models will be deprecated soon, as per this schedule [deprecations](https://ai.google.dev/gemini-api/docs/deprecations#imagen-models).

I would recommend migrating to the Gemini 3.x pro image models beforehand, while using the Google GenAI SDK.

Thank you

cc @ilayaperumalg @sdeleuze

OreOreDa commented:
@ddobrin  @ilayaperumalg @sdeleuze 

My new [PR](https://github.com/spring-projects/spring-ai/pull/6408) for Google GenAI image models
I integrated the latest refacto, and the latest GenAI sdk updates - tested with "gemini-2.5-flash-image" so far
Can you have a look at it, please ?

Edit: I re created the PR after rebase to the 2.0.0 - congrats for the release, Spring AI is awesome !

sdeleuze commented:
Closing the PR in favor of #6408.
